### PR TITLE
Better unification between the dye, wool, and bed mods.

### PR DESCRIPTION
**Pull request by treytomes**
Changes to the dye mod:
1. All dyes are generated dynamically from the dye.dyes table, which I added html color values to.  More dyes can be created simply by adding to this table.
2. All colors are now located in a single "color_"..<color name> group.  It may simply be my own ignorance, but having 3 different color sets seemed to be needlessly complicating color usage.
3. Replaced the texture set with 2 textures: dye_base and dye_overlay.  These are combined into a single texture using the html colors and some rgb<-->hsv helper functions that I added.

Changes to the wool mod:
1. Wool blocks are generated dynamically from the dyes.dye table.
2. Wool textures have been replaced by a wool_base.png and wool_overlay.png, with wool_base.png colorized by the dye html color.

Changes to the beds mod:
1. The only difference between PA beds and minetest-game beds seemed to be the multiple color selection, so I added that functionality to this mod.
2. A set of _overlay.png textures have been added to colorize the beds.
3. Simple and fancy beds are generated based on the dyes.dye table.

That's all for today.  Please let me know what you think.

### CODE REVIEW

**Comment by paramat** [comment]
A possible problem is that all texture packs will be borken for dyes, wools and beds.

**Comment by treytomes** [comment]
That makes sense.  Any texture packs that override the default dye, wool, or bed images would be broken.  On the other hand, generating the textures this way means that it would not take nearly as much work to rebuild them.

Also, is there a reason why there are 3 different color groups, or is it just legacy code that no one has taken the time to remove yet?  If there is a reason to it that still applies, maybe I can find a way to handle that procedurally.

**Comment by paramat** [comment]
Knowing MT probably just messy development :)

File: minetest.conf [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#
```
Comment (sofar) [inline minetest.conf]: unrelated commit, please remove - something like this needs to be a separate pull request.

File: mods/dye/init.lua [hunk 2]
```diff
@@ -1,71 +1,44 @@
...
-	{"cyan",       "Cyan dye",       {dye=1, basecolor_cyan=1,    excolor_cyan=1,       unicolor_cyan=1}},
-	{"dark_green", "Dark green dye", {dye=1, basecolor_green=1,   excolor_green=1,      unicolor_dark_green=1}},
-	{"green",      "Green dye",      {dye=1, basecolor_green=1,   excolor_green=1,      unicolor_green=1}},
-	{"yellow",     "Yellow dye",     {dye=1, basecolor_yellow=1,  excolor_yellow=1,     unicolor_yellow=1}},
-	{"brown",      "Brown dye",      {dye=1, basecolor_brown=1,   excolor_orange=1,     unicolor_dark_orange=1}},
-	{"orange",     "Orange dye",     {dye=1, basecolor_orange=1,  excolor_orange=1,     unicolor_orange=1}},
-	{"red",        "Red dye",        {dye=1, basecolor_red=1,     excolor_red=1,        unicolor_red=1}},
-	{"magenta",    "Magenta dye",    {dye=1, basecolor_magenta=1, excolor_red_violet=1, unicolor_red_violet=1}},
-	{"pink",       "Pink dye",       {dye=1, basecolor_red=1,     excolor_red=1,        unicolor_light_red=1}},
+	white		= { name = "White", html = "#FFFFFF" },
```
Comment (sofar) [inline mods/dye/init.lua]: no space after '{' and before '}'

File: mods/dye/init.lua [hunk 3]
```diff
@@ -1,71 +1,44 @@
...
+	black		= { name = "Black", html = "#000000" },
+	violet		= { name = "Violet", html = "#8000FF" },
+	blue		= { name = "Blue", html = "#0000FF" },
+	cyan		= { name = "Cyan", html = "#00FFFF" },
+	dark_green	= { name = "Dark Green", html = "#007F00" },
+	green		= { name = "Green", html = "#00FF00" },
+	yellow		= { name = "Yellow", html = "#FFFF00" },
+	brown		= { name = "Brown", html = "#854C30" },
+	orange		= { name = "Orange", html = "#FF7F00" },
+	red			= { name = "Red", html = "#FF0000" },
```
Comment (sofar) [inline mods/dye/init.lua]: tab too much here.

File: mods/dye/init.lua [hunk 4]
```diff
@@ -1,71 +1,44 @@
...
-	minetest.register_craftitem(item_name, {
-		inventory_image = item_image,
-		description = description,
-		groups = groups
+for key, value in pairs(dye.dyes) do
+	local itemName = "dye:"..key
+
+	minetest.register_craftitem(itemName, {
+		inventory_image = "dye_"..key..".png",
+		description = value.name,
```
Comment (sofar) [inline mods/dye/init.lua]: Add `" dye"` to the description. The item isn't the color green, it's `"Green dye"`

File: mods/dye/init.lua [hunk 5]
```diff
@@ -1,71 +1,44 @@
...
-	local description = row[2]
-	local groups = row[3]
-	local item_name = "dye:" .. name
-	local item_image = "dye_" .. name .. ".png"
-	minetest.register_craftitem(item_name, {
-		inventory_image = item_image,
-		description = description,
-		groups = groups
+for key, value in pairs(dye.dyes) do
+	local itemName = "dye:"..key
```
Comment (sofar) [inline mods/dye/init.lua]: say no to camelCaps!

File: mods/dye/init.lua [hunk 6]
```diff
@@ -1,71 +1,44 @@
...
-		output = item_name .. " 4",
-		recipe = {"group:flower,color_" .. name},
+		output = itemName .. " 4",
+		recipe = {"group:flower,color_"..key},
 	})
 end
 
--- Manually add coal->black dye
+-- Manually add coal->black dye.
+-- TODO: Is this still desirable?
```
Comment (sofar) [inline mods/dye/init.lua]: It's still needed, so this TODO should not be added.

File: mods/wool/init.lua [hunk 7]
```diff
@@ -2,46 +2,68 @@
...
-
-for i = 1, #dyes do
-	local name, desc, craft_color_group = unpack(dyes[i])
-
-	minetest.register_node("wool:" .. name, {
-		description = desc .. " Wool",
-		tiles = {"wool_" .. name .. ".png"},
+for key, value in pairs(dye.dyes) do
+	local color = key
+	local colorDisplayName = value.name
```
Comment (sofar) [inline mods/wool/init.lua]: No camels, please

File: mods/wool/init.lua [hunk 8]
```diff
@@ -2,46 +2,68 @@
...
+	local htmlColor = value.html.."A0"
+	local craft_color_group = "color_"..color
+
+	minetest.register_node("wool:"..color, {
+		description = colorDisplayName.." Wool",
+		tiles = {"wool.png^[colorize:"..htmlColor},
```
Comment (sofar) [inline mods/wool/init.lua]: if you'd use `wool_white.png` to colorize, you wouldn't break any texture pack.

File: mods/wool/init.lua [hunk 9]
```diff
@@ -2,46 +2,68 @@
...
 		is_ground_content = false,
-		groups = {snappy = 2, choppy = 2, oddly_breakable_by_hand = 3,
-				flammable = 3, wool = 1},
+		groups = {snappy = 2, choppy = 2, oddly_breakable_by_hand = 3, flammable = 3, wool = 1},
 		sounds = default.node_sound_defaults(),
 	})
 
-	minetest.register_craft{
```
Comment (sofar) [inline mods/wool/init.lua]: how did this ever work? o_O

File: mods/wool/init.lua [hunk 10]
```diff
@@ -2,46 +2,68 @@
...
 		type = "shapeless",
-		output = "wool:" .. name,
-		recipe = {"group:dye," .. craft_color_group, "group:wool"},
-	}
+		output = "wool:"..color,
+		recipe = {"group:dye,"..craft_color_group, "group:wool"},
+	})
 end
 
+--local dyes = {
```
Comment (sofar) [inline mods/wool/init.lua]: delete all this instead of commenting it out.

File: mods/wool/init.lua [hunk 11]
```diff
@@ -2,46 +2,68 @@
...
+--		type = "shapeless",
+--		output = "wool:" .. name,
+--		recipe = {"group:dye," .. craft_color_group, "group:wool"},
+--	}
+--end
+
 
 -- legacy
 
 -- Backwards compatibility with jordach's 16-color wool mod
```
Comment (sofar) [inline mods/wool/init.lua]: you're deleting the aliases, but not the comment? that doesn't make sense.

File: mods/beds/functions.lua [hunk 12]
```diff
@@ -1,8 +1,9 @@
 local pi = math.pi
 local player_in_bed = 0
-local is_sp = minetest.is_singleplayer()
+local _isSinglePlayer = minetest.is_singleplayer()
```
Comment (sofar) [inline mods/beds/functions.lua]: Sorry, camels belong in the desert, not in code :)

File: mods/beds/functions.lua [hunk 13]
```diff
@@ -137,6 +138,9 @@ function beds.on_rightclick(pos, player)
 	local ppos = player:getpos()
 	local tod = minetest.get_timeofday()
 
+	beds.set_spawn(name)
```
Comment (sofar) [inline mods/beds/functions.lua]: unrelated functionality change! please remove and submit a *separate* pr for this change.

File: mods/beds/functions.lua [hunk 14]
```diff
@@ -171,21 +175,26 @@ function beds.on_rightclick(pos, player)
...
+
+-- Only register respawn callback if respawn enabled.
 if enable_respawn then
-	-- respawn player at bed if enabled and valid position is found
+	-- Respawn player at bed if enabled and valid position is found.
 	minetest.register_on_respawnplayer(function(player)
 		local name = player:get_player_name()
-		local pos = beds.spawn[name]
-		if pos then
+		local pos = beds.spawn[name] -- Get the position of the player's bed.
```
Comment (sofar) [inline mods/beds/functions.lua]: Too detailed comments. The function of these lines is not hard to understand. The comments take it way too far.

File: mods/beds/functions.lua [hunk 15]
```diff
@@ -171,21 +175,26 @@ function beds.on_rightclick(pos, player)
...
+		if pos then -- Do this if the player has a bed.
 			player:setpos(pos)
 			return true
 		end
 	end)
 end
 
+-- Called when the player leaves the game.
```
Comment (sofar) [inline mods/beds/functions.lua]: again, way too much comments.

File: mods/beds/functions.lua [hunk 16]
```diff
@@ -201,6 +210,8 @@ minetest.register_on_leaveplayer(function(player)
 	end
 end)
 
+-- Called when a button is pressed in player's inventory form, newest functions are called first. If function returns true, remaining functions are not called.
```
Comment (sofar) [inline mods/beds/functions.lua]: too much documentation.

File: mods/beds/spawns.lua [hunk 17]
```diff
@@ -54,10 +54,20 @@ function beds.set_spawns()
...
-		-- but don't change spawn location if borrowing a bed
+		-- ...but don't change spawn location if borrowing a bed.
 		if not minetest.is_protected(p, name) then
 			beds.spawn[name] = p
 		end
 	end
 	beds.save_spawns()
 end
+
+function beds.set_spawn(playerName)
```
Comment (sofar) [inline mods/beds/spawns.lua]: again, remove this from the dye changes. This should be a separate PR. You also didn't document this change and since it's an exported function, API documentation is missing.

File: mods/dye/init.lua [hunk 18]
```diff
@@ -20,13 +20,104 @@ dye.dyes = {
...
 
--- Define items
+----
+-- Helper functions.
+----
+
+--[[
+	Convert an html color in the format of #RRGGBB into red, green, and blue values on the range of [0..255].
+]]
+function hex2rgb(hex)
```
Comment (sofar) [inline mods/dye/init.lua]: API / public functions. This should be clear that these are public and require documentation, as well as put in a reasonable namespace (e.g. dye.hex2rgb())

File: mods/dye/init.lua [hunk 19]
```diff
@@ -20,13 +20,104 @@ dye.dyes = {
...
+    hex = hex:gsub("#","")
```
Comment (sofar) [inline mods/dye/init.lua]: tabs vs spaces

File: mods/dye/init.lua [hunk 20]
```diff
@@ -20,13 +20,104 @@ dye.dyes = {
...
+	local h, s, v = max, max, max
+
+	local d = max - min;
+	if max == 0 then
+		s = 0
+	else
+		s = d / max
+	end
+
+	if(max == min) then
```
Comment (sofar) [inline mods/dye/init.lua]: space after `if`

File: mods/dye/init.lua [hunk 21]
```diff
@@ -20,13 +20,104 @@ dye.dyes = {
...
+	local r = red / 255
+	local g = green / 255
+	local b = blue / 255
+	local max = math.max(r, g, b)
+	local min = math.min(r, g, b)
+	local h, s, v = max, max, max
+
+	local d = max - min;
+	if max == 0 then
+		s = 0
```
Comment (sofar) [inline mods/dye/init.lua]: if `max == 0`, then `s` already is 0.

File: mods/dye/init.lua [hunk 22]
```diff
@@ -20,13 +20,104 @@ dye.dyes = {
...
+]]
+function hsvToRgb(h, s, v)
+	local i = math.floor(h * 6);
+	local f = h * 6 - i;
+	local p = v * (1 - s);
+	local q = v * (1 - f * s);
+	local t = v * (1 - (1 - f) * s);
+
+	local r, g, b;
+	local switch = i % 6
```
Comment (sofar) [inline mods/dye/init.lua]: you could probably use a table with functions to eliminate 6 branches in exchange for a table lookup and a function call.

File: mods/dye/init.lua [hunk 23]
```diff
@@ -20,13 +20,104 @@ dye.dyes = {
...
+end
+
+----
+-- Define items.
+----
 
 for key, value in pairs(dye.dyes) do
 	local itemName = "dye:"..key
 
+	minetest.log("info", "Darkened dye: "..value.html.." to: "..darken(value.html))
```
Comment (sofar) [inline mods/dye/init.lua]: darkening? where did that come from?

**Comment by sofar** [comment]
This is a lot of code all at once.

I would strongly suggest focusing just on the dye part first. Then do wool, and last beds, in separate PRs. I can see dye and wool changes be in the same PR (2 commits), but not beds, since that changes beds significantly and adds 28 new beds to the game.

File: mods/dye/init.lua (hunk 18 above)
Comment (treytomes) [inline mods/dye/init.lua]: I actually meant for this to be hex2rgb, rgb2hex, rgbToHsv, hsvToRgb, and darken to be local functions.  I can change these to local, then remove the camel casing.

File: mods/dye/init.lua (hunk 21 above)
Comment (treytomes) [inline mods/dye/init.lua]: That's what I get for cutting-and-pasting code.  I can standardize this as well.

File: mods/dye/init.lua (hunk 22 above)
Comment (treytomes) [inline mods/dye/init.lua]: I'll give that a try.

File: mods/wool/init.lua (hunk 9 above)
Comment (treytomes) [inline mods/wool/init.lua]: I think this is a case of Lua not demanding the use of parenthesis around function parameters.  The lack made the code harder to read though.

File: mods/wool/init.lua (hunk 11 above)
Comment (treytomes) [inline mods/wool/init.lua]: I left the aliases commented out in case someone wanted them back in later, but I suppose that's what source control is for.

**Comment by treytomes** [comment]
@sofar Thank you for taking the time to review my code.  I will submit a different set of pull requests once I get these changes into place.

File: mods/wool/init.lua (hunk 9 above)
Comment (numberZero) [inline mods/wool/init.lua]: Lua supports `func"some string"` as well ;) http://www.lua.org/manual/5.1/manual.html#2.5.8
But these forms look ugly IMO

Comment (treytomes) [inline mods/wool/init.lua]: Yes, I've read about this.  It feels like a bad idea for most of the cases that I can think of.

**Comment by sofar** [comment]
Just a note: when I send a lot of review comments, it's usually because I think it's worth doing and support the effort! :)

File: mods/wool/init.lua (hunk 7 above)
Comment (treytomes) [inline mods/wool/init.lua]: The camels are being evicted as we speak.

**Comment by treytomes** [comment]
Thanks!

**Comment by Ekdohibs** [comment]
What about using the new `colored_facedir` paramtype2 so that the number of beds that needs to be added is smaller?

**Comment by paramat** [comment]
Ekdohibs param2 coloured nodes are not distinct in inventory yet (i'm sure you know).

**Comment by sofar** [comment]
I really think this PR would get along quicker if it avoided the bed patch entirely...

**Comment by treytomes** [comment]
I'm planning to submit 2 pull requests for the dye and wool mods later today.  I'll either need to figure out how to submit selected files from git, or rebuild my local repository (which wouldn't really be that hard).

The bed color change seemed like a natural extension of the existing mod, but the feature to only set the spawn point may or may not belong in a separate mod.  At the very least I feel like I should attach it to a configuration value.

**Comment by treytomes** [comment]
I will probably split the bed changes into 2 pull requests, and my changes represent 2 distinct features, then let you all decide whether they are worth keeping or not.

**Comment by paramat** [comment]
Also i think best to use existing multiple textures, to not break texture packs, for now, then using texture colourisation can be a separate PR since it's controversial.
The issue of using texture colourisaton is that colourising a single base image doesn't have as much flexibility as separate textures for each wool, there is a better result when tweaking brightness, contrast etc. for each texture. This issue doesn't affect dyes much because their appearence isn't very important.

**Comment by paramat** [comment]
:-1: I disapprove of colourising wool textures, as this removes the ability to carefully customise each wool texture independently. I also do not like how the dye colour is the colouriser, it may seem obvious to do but the colour of a dye item is not necessarliy ideal for the colour of the wool node, hue yes but saturation and value may be better tuned for each wool.
This also breaks texture packs and forces them to use a much less flexible way to texture wool.

So i feel colourisation and hardware colourisation should be used for newly added items only. Coloured beds are new so that would be ok.

**Comment by paramat** [comment]
See new PR.
